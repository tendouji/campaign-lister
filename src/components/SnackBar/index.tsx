import React, { Component } from 'react';
import styled from 'styled-components';
import { colors, layout, styles } from "../../constants/ui";
import { config } from "../../constants/config";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTimesCircle } from "@fortawesome/free-solid-svg-icons/faTimesCircle";


type SnackBarProps = {
    show: false,
    message: '',
    close: () => void,
}

class SnackBar extends Component<SnackBarProps, {}> {
    private snackBarTimer: ReturnType<typeof setTimeout> = 0;

    constructor(props: any) {
        super(props);
        this.closeHandler = this.closeHandler.bind(this);
    }

    closeHandler = () => {
        const { close } = this.props;
        close();
    };

    componentDidMount() { }

    componentDidUpdate(prevProps: SnackBarProps) {
        const { show } = this.props;

        if(prevProps.show !== show) {
            if(!!show) {
                if(!!this.snackBarTimer) {
                    clearTimeout(this.snackBarTimer);
                }
                this.snackBarTimer = setTimeout(() => {
                    this.closeHandler();
                }, config.snackBarLifeSpan);
            }
        }
    }

    render() {
        const { show, message } = this.props;

        return (
            <SnackBarWrapper
                className={[
                    'snackbar',
                    !!show ? ` show` : '',
                ].join('')}
            >
                <div className="content">{message}</div>
                <div className="close" onClick={this.closeHandler}>
                    <FontAwesomeIcon icon={faTimesCircle} size={'lg'} />
                </div>
            </SnackBarWrapper>
        )
    }
}


const closeSize = '20px';
const transitionTime = config.snackBarTransition;
const SnackBarWrapper = styled.div`
    display: flex;
    position: fixed;
    top: -200px;
    right: ${layout.standardGap};
    max-width: calc(100% - 2 * ${layout.standardGap});
    min-width: 300px;
    padding: ${layout.smallGap} ${layout.standardGap};
    border-radius: ${layout.borderRadius};
    box-sizing: border-box;
    color: ${colors.white};
    background-color: ${colors.black};
    justify-content: space-between;
    transition: top ${transitionTime * 2}ms;
    
    &.show {
        top: ${layout.standardGap};
        transition: top ${transitionTime}ms;
    }
    
    & > .content {
        width: calc(100% - ${closeSize});
        margin-right: ${layout.standardGap};
        line-height: ${closeSize};
        font-size: ${layout.fontSizeStandard};
        font-weight: 400;
        ${styles.textOverflowStyle}
    }
    
    & > .close {
        height: ${closeSize};
        width: ${closeSize};
        
        & * {
            cursor: pointer;
        }
    }
`;


export default SnackBar;
