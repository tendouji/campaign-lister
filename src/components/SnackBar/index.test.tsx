import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import SnackBar from './';


describe('<SnackBar />', () => {
    const props = {
        show: true,
        message: 'Test Message',
    };

    it('renders without crashing with props', () => {
        const wrapper = shallow(<SnackBar {...props} />);
        expect(wrapper.hasClass('show')).toEqual(true);
        expect(toJson(wrapper)).toMatchSnapshot()
    });

    it('move out of screen when close button is clicked', () => {
        const wrapper = shallow(<SnackBar {...props} />);
        wrapper.setProps({
            close: () => wrapper.setProps({ show: false })
        });

        const closeBtn = wrapper.find('.close');
        closeBtn.simulate('click');

        expect(wrapper.hasClass('show')).toEqual(false);
    });
});