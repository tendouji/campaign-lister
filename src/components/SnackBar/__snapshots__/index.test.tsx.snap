// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SnackBar /> renders without crashing with props 1`] = `
<styled.div
  className="snackbar show"
>
  <div
    className="content"
  >
    Test Message
  </div>
  <div
    className="close"
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f057",
            "M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z",
          ],
          "iconName": "times-circle",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size="lg"
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </div>
</styled.div>
`;
